data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java

```diff
@@ -3,300 +3,255 @@
-            accessToken = csb.getAccessToken();
-        } else if(csb.getTokenProvider() != null){
-            authenticationType = AuthenticationType.AAD_ACCESS_TOKEN_PROVIDER;
-            tokenProvider = csb.getTokenProvider();
-        } else {
-            authenticationType = AuthenticationType.AAD_DEVICE_LOGIN;
+        scope = String.format("%s/%s", clusterUrl, ".default");
+        scopes = new HashSet<>();
+        scopes.add(scope);
+
+        aadAuthorityUrl = determineAadAuthorityUrl(csb.getAuthorityId());
+
+        try {
+            if (StringUtils.isNotBlank(csb.getApplicationClientId()) && StringUtils.isNotBlank(csb.getApplicationKey())) {
+                clientId = csb.getApplicationClientId();
+                clientSecret = ClientCredentialFactory.createFromSecret(csb.getApplicationKey());
+                confidentialClientApplication = ConfidentialClientApplication.builder(clientId, clientSecret).authority(aadAuthorityUrl).build();
+                authenticationType = AuthenticationType.AAD_APPLICATION_KEY;
+            } else if (csb.getX509Certificate() != null && csb.getPrivateKey() != null && StringUtils.isNotBlank(csb.getApplicationClientId())) {
+                clientCertificate = ClientCredentialFactory.createFromCertificate(csb.getPrivateKey(), csb.getX509Certificate());
+                applicationClientId = csb.getApplicationClientId();
+                confidentialClientApplication = ConfidentialClientApplication.builder(applicationClientId, clientCertificate).authority(aadAuthorityUrl).validateAuthority(false).build();
+                authenticationType = AuthenticationType.AAD_APPLICATION_CERTIFICATE;
+            } else if (StringUtils.isNotBlank(csb.getAccessToken())) {
+                accessToken = csb.getAccessToken();
+                authenticationType = AuthenticationType.AAD_ACCESS_TOKEN;
+            } else if (csb.getTokenProvider() != null) {
+                tokenProvider = csb.getTokenProvider();
+                authenticationType = AuthenticationType.AAD_ACCESS_TOKEN_PROVIDER;
+            } else {
+                if (StringUtils.isNotBlank(csb.getUserUsernameHint())) {
+                    userUsername = csb.getUserUsernameHint();
+                }
+                redirectUri = new URI("http://localhost");
+                publicClientApplication = PublicClientApplication.builder(CLIENT_ID).authority(aadAuthorityUrl).build();
+                authenticationType = AuthenticationType.AAD_USER_PROMPT;
+            }
+        } catch (MalformedURLException e) {
+            throw new URISyntaxException(aadAuthorityUrl, ERROR_INVALID_AUTHORITY_URL);
         }
+    }
 
-        // Set the AAD Authority URI
-        String aadAuthorityId = (csb.getAuthorityId() == null ? DEFAULT_AAD_TENANT : csb.getAuthorityId());
+    private String determineAadAuthorityUrl(String aadAuthorityUrl) {
+        String aadAuthorityId = (aadAuthorityUrl == null ? ORGANIZATION_URI_SUFFIX : aadAuthorityUrl);
```

Comment:
Is it actually the authority ID you pass in the parameter?
Why is authUrl consider authId?

Reply:
Good point, this is quite confusing.